models/issue_dependency.go

```diff
@@ -0,0 +1,154 @@
+// Use of this source code is governed by a MIT-style
+// license that can be found in the LICENSE file.
+
+package models
+
+import (
+	"time"
+)
+
+// IssueDependency is connection request for receiving issue notification.
+type IssueDependency struct {
+	ID           int64     `xorm:"pk autoincr"`
+	UserID       int64     `xorm:"UNIQUE(watch) NOT NULL"`
+	IssueID      int64     `xorm:"UNIQUE(watch) NOT NULL"`
+	DependencyID int64     `xorm:"UNIQUE(watch) NOT NULL"`
+	Created      time.Time `xorm:"-"`
+	CreatedUnix  int64     `xorm:"INDEX created"`
+	Updated      time.Time `xorm:"-"`
+	UpdatedUnix  int64     `xorm:"updated"`
+}
+
+// Define Dependency Type Constants
+const (
+	DependencyTypeBlockedBy int64 = 1
+	DependencyTypeBlocking  int64 = 2
+)
+
+// CreateIssueDependency creates a new dependency for an issue
+func CreateIssueDependency(user *User, issue, dep *Issue) (exists bool, err error) {
+	sess := x.NewSession()
+
+	// Check if it aleready exists
+	exists, err = issueDepExists(x, issue.ID, dep.ID)
+	if err != nil {
+		return exists, err
+	}
+
+	// If it not exists, create it, otherwise show an error message
+	if !exists {
+		newIssueDependency := &IssueDependency{
+			UserID:       user.ID,
+			IssueID:      issue.ID,
+			DependencyID: dep.ID,
+		}
+
+		if _, err := x.Insert(newIssueDependency); err != nil {
+			return exists, err
+		}
+
+		// Add comment referencing the new dependency
+		_, err = createIssueDependencyComment(sess, user, issue, dep, true)
+
+		if err != nil {
+			return exists, err
+		}
+
+		// Create a new comment for the dependent issue
+		_, err = createIssueDependencyComment(sess, user, dep, issue, true)
+
+		if err != nil {
+			return exists, err
+		}
+	}
+	return exists, nil
+}
+
+// RemoveIssueDependency removes a dependency from an issue
+func RemoveIssueDependency(user *User, issue *Issue, dep *Issue, depType int64) (err error) {
+	sess := x.NewSession()
+
+	// Check if it exists
+	exists, err := issueDepExists(x, issue.ID, dep.ID)
+	if err != nil {
+		return err
+	}
+
+	// If it exists, remove it, otherwise show an error message
+	if exists {
+
+		if depType == DependencyTypeBlockedBy {
+			_, err := x.Delete(&IssueDependency{IssueID: issue.ID, DependencyID: dep.ID})
+			if err != nil {
+				return err
+			}
+		}
+
+		if depType == DependencyTypeBlocking {
+			_, err := x.Delete(&IssueDependency{IssueID: dep.ID, DependencyID: issue.ID})
+			if err != nil {
+				return err
+			}
+		}
+
+		// Add comment referencing the removed dependency
+		_, err = createIssueDependencyComment(sess, user, issue, dep, false)
+
+		if err != nil {
+			return err
+		}
+
+		// Create a new comment for the dependent issue
+		_, err = createIssueDependencyComment(sess, user, dep, issue, false)
+
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// Check if the dependency already exists
+func issueDepExists(e Engine, issueID int64, depID int64) (exists bool, err error) {
```

Comment:
I don't like names `issueID` and `depID` because you are switching their meaning inside function.

Reply:
How do you mean this?

Reply:
First `issueID` means `issueID` and second time `issueID` means `depID`. I think both arguments are equivalent because you are trying to find dependency between them not matter which way. Variable names should reflect this fact (eg. `issue1ID, issue2ID int64`).

Reply:
You're right, I try to find all dependencies, no matter of direction.
But: `issueID` is the issue we're currently on, `depID` is the other issue we want to look for. I think it would be more confunsing to rename the variables to `issue1ID` etc. I would prefer keeping it the way it is.

To be more clear, an example:
When I'm on issue #1 and want to add a new dependency of #2 i send a request to the backend saying something like "Hi, I'm issue #1 and want to add #2 as my dependent issue.". The dependend issue (#2) becomes `depID`, the issue we're on becomes `isuueID`.

Reply:
I am ok with that as long as `issueDepExists` stays unexported and is used only for your mentioned scenario.

I was thinking about it in more general way. Like using this function for checking (from any place of code) any dependency between two issues.

Reply:
Ok, cool.

I think to be more general, it  would need to return a struct with both issues, but i dont see a usecase for this (yet).